app/src/main/java/com/onewelcome/showcaseapp/feature/userauthentication/pinauthentication/PinAuthenticationViewModel.kt

```diff
@@ -0,0 +1,141 @@
+  private val getUserProfilesUseCase: GetUserProfilesUseCase,
+  private val pinAuthenticationRequestHandler: PinAuthenticationRequestHandler,
+  private val getAuthenticatedUserProfileUseCase: GetAuthenticatedUserProfileUseCase
+) : ViewModel() {
+  var uiState by mutableStateOf(State())
+    private set
+
+  private val _navigationEvents = Channel<NavigationEvent>(Channel.BUFFERED)
+  val navigationEvents = _navigationEvents.receiveAsFlow()
+
+  fun onEvent(event: UiEvent) {
+    when (event) {
+      is UiEvent.StartPinAuthentication -> startPinAuthentication()
+      is UiEvent.CancelAuthentication -> cancelAuthentication()
+      is UiEvent.UpdateSelectedUserProfile -> updateSelectedUserProfile(event)
+      is UiEvent.LoadData -> loadData()
+    }
+  }
+
+  private fun updateSelectedUserProfile(event: UiEvent.UpdateSelectedUserProfile) {
+    uiState = uiState.copy(selectedUserProfile = event.userProfile)
+  }
+
+  private fun loadData() {
+    viewModelScope.launch {
+      updateIsSdkInitialized()
+      updateUserProfiles()
+      updateAuthenticatedProfile()
+      updateAuthenticateButton()
+    }
+  }
+
+  private suspend fun updateAuthenticatedProfile() {
+    getAuthenticatedUserProfileUseCase.execute()
+      .onSuccess { uiState = uiState.copy(authenticatedUserProfile = it) }
+      .onFailure { uiState = uiState.copy(authenticatedUserProfile = null) }
+  }
+
+  private fun updateIsSdkInitialized() {
+    isSdkInitializedUseCase.execute().let { uiState = uiState.copy(isSdkInitialized = it) }
+  }
+
+  private fun updateAuthenticateButton() {
+    uiState = uiState.copy(isAuthenticateButtonEnabled = uiState.isSdkInitialized && uiState.selectedUserProfile != null)
+  }
+
+  private suspend fun updateUserProfiles() {
+    getUserProfilesUseCase.execute()
+      .onSuccess { uiState = uiState.copy(userProfiles = it, selectedUserProfile = it.firstOrNull()) }
+      .onFailure { uiState = uiState.copy(userProfiles = emptySet()) }
+  }
+
+  private fun cancelAuthentication() {
+    pinAuthenticationRequestHandler.pinCallback?.denyAuthenticationRequest()
+  }
+
+  private fun startPinAuthentication() {
+    authenticateUser()
+    listenForPinScreenNavigationEvent()
+  }
+
+  private fun authenticateUser() {
+    uiState.selectedUserProfile?.let { selectedUserProfile ->
+      viewModelScope.launch {
+        getRegisteredAuthenticatorsUseCase.execute(selectedUserProfile)
+          .onSuccess {
+            runCatching {
+              val pinAuthenticator = it.first { it.type == OneginiAuthenticator.Type.PIN }
+              pinAuthenticationUseCase.execute(selectedUserProfile, pinAuthenticator)
+                .onSuccess { uiState = uiState.copy(result = Ok(it)) }
+                .onFailure { uiState = uiState.copy(result = Err(it)) }
+            }.onFailure { uiState = uiState.copy(result = Err(it)) }
+          }
+          .onFailure { uiState = uiState.copy(result = Err(it)) }
```

Comment:
This could be flatten to something like this:
```kotlin
getRegisteredAuthenticatorsUseCase.execute(selectedUserProfile)
          .map { it.first { it.type == OneginiAuthenticator.Type.PIN } }
          .flatMap { pinAuthenticationUseCase.execute(selectedUserProfile, it) }
          .onSuccess { uiState = uiState.copy(result = Ok(it)) }
          .onFailure { uiState = uiState.copy(result = Err(it)) }
```